lib/ramble/ramble/context.py

```diff
@@ -0,0 +1,142 @@
+
+
+class Context(object):
+    """Class to represent a context
+
+    This class contains variable definitions to store any individual context
+    (such as application, workload, or experiment) and logic to merge in
+    additional contexts by order of precedence."""
+
+    def __init__(self):
+        """Constructor for a Context
+
+        Create a Context object, which holds context attributes.
+        """
+        self.env_variables = []
+        self.variables = syaml.syaml_dict()
+        self.internals = {}
+        self.templates = None
+        self.chained_experiments = []
+        self.modifiers = []
+        self.context_names = None
+        self.exclude = {}
+        self.zips = {}
+        self.matrices = []
+        self.is_template = False
+
+    def merge_context(self, in_context):
+        """Merges another Context into this Context."""
+
+        internal_sections = [namespace.custom_executables,
+                             namespace.executables]
+
+        if in_context.variables:
+            self.variables.update(in_context.variables)
+        if in_context.env_variables:
+            self.env_variables.append(in_context.env_variables)
+        if in_context.internals:
+            for internal_section in internal_sections:
+                if internal_section in in_context.internals:
+                    if isinstance(in_context.internals[internal_section], dict):
+                        if internal_section not in self.internals:
+                            self.internals[internal_section] = {}
+                        section_dict = in_context.internals[internal_section]
+                        for key, val in section_dict.items():
+                            self.internals[internal_section][key] = val
+                    elif isinstance(in_context.internals[internal_section], list):
+                        if internal_section not in self.internals:
+                            self.internals[internal_section] = []
+                        self.internals[internal_section].extend(
+                            in_context.internals[internal_section])
+                    else:
+                        self.internals[internal_section] = \
+                            in_context.internals[internal_section]
+        if in_context.chained_experiments:
+            for chained_exp in in_context.chained_experiments:
+                self.chained_experiments.append(chained_exp.copy())
+        if in_context.modifiers:
+            for modifier in in_context.modifiers:
+                self.modifiers.append(modifier.copy())
+        if in_context.templates is not None:
+            self.is_template = in_context.templates
+        if in_context.exclude:
+            self.exclude = in_context.exclude
+        if in_context.zips:
+            self.zips = in_context.zips
+        if in_context.matrices:
+            self.matrices = in_context.matrices
+
+
+def from_dict(context_name, in_dict):
```

Comment:
I think this should either be a member function of `Context` (where it should set the context from a dictionary), i.e.:
```
new_context = Context()
new_context.from_dict(in_dict)
```

Or, the name should be more like `create_context_from_dict`.

Reply:
Great, I've updated the name. I might be overthinking this, but my reasoning was that having this as a non-member function would ensure that updates to an existing instance go through the 'merge_context' logic. That way, a context 'from_dict' would start with a clean instance and avoid any potential conflict with the 'from_dict' method being invoked after vars have already been set and either not overwriting or not properly merging the two contexts.